libs/utils/website/src/lib/async-session.provider.tsx

```diff
@@ -0,0 +1,75 @@
+import {User, UserSession} from '@wepublish/website/api'
+import {AuthTokenStorageKey, SessionTokenContext} from '@wepublish/authentication/website'
+import {useMeLazyQuery} from '@wepublish/website/api'
+import {memo, PropsWithChildren, useCallback, useEffect, useRef, useState} from 'react'
+
+export const AsyncSessionProvider = memo<PropsWithChildren<{sessionToken: UserSession | null}>>(
+  function SessionProvider({sessionToken, children}) {
+    const [token, setToken] = useState<typeof sessionToken>(sessionToken)
+    const [user, setUser] = useState<User | null>(null)
+    const initialSetupDone = useRef(false)
+
+    const [getMe] = useMeLazyQuery({
+      onCompleted(data) {
+        setUser((data.me as User) ?? null)
+      }
+    })
+
+    const setCookieAndToken = useCallback(
+      async (newToken: UserSession | null) => {
+        setToken(newToken)
+
+        if (newToken) {
+          sessionStorage.setItem(AuthTokenStorageKey, JSON.stringify(newToken))
+          await fetch('/api/cookie', {
+            method: 'POST',
+            body: JSON.stringify(newToken),
+            headers: {
+              'Content-Type': 'application/json'
+            }
+          })
+          await getMe()
+        } else {
+          sessionStorage.removeItem(AuthTokenStorageKey)
+          setUser(null)
+          await fetch('/api/cookie', {
+            method: 'DELETE'
+          })
+        }
+      },
+      [getMe]
+    )
+
+    const setupInitialSessionToken = useCallback(async () => {
+      if (sessionToken) {
```

Comment:
This is when the token comes from `getInitialProps` or `getServerSideProps`, then we don't need to fetch it from the API